# Data Visualization on Space Missions from 1957

## Introduction 

This DataSet was scraped from https://nextspaceflight.com/launches/past/?page=1 and includes all the space missions since the beginning of Space Race (1957).

## 1. Import Libraries

In [52]:
import warnings

warnings.filterwarnings('ignore')

In [53]:
ches

## 2. Load the data

In [54]:
space = pd.read_csv("datasets_828921_1417131_Space_Corrected.csv")
space.head(10)

Unnamed: 0  Unnamed: 0.1 Company Name  \
0           0             0       SpaceX   
1           1             1         CASC   
2           2             2       SpaceX   
3           3             3    Roscosmos   
4           4             4          ULA   
5           5             5         CASC   
6           6             6    Roscosmos   
7           7             7         CASC   
8           8             8       SpaceX   
9           9             9         JAXA   

                                            Location  \
0         LC-39A, Kennedy Space Center, Florida, USA   
1  Site 9401 (SLS-2), Jiuquan Satellite Launch Ce...   
2                      Pad A, Boca Chica, Texas, USA   
3       Site 200/39, Baikonur Cosmodrome, Kazakhstan   
4           SLC-41, Cape Canaveral AFS, Florida, USA   
5       LC-9, Taiyuan Satellite Launch Center, China   
6         Site 31/6, Baikonur Cosmodrome, Kazakhstan   
7    LC-101, Wenchang Satellite Launch Center, China   
8           SLC-40, Cape Canaveral AFS, Florida, USA   
9             LA-Y1, Tanegashima Space Center, Japan   

                        Datum  \
0  Fri Aug 07, 2020 05:12 UTC   
1  Thu Aug 06, 2020 04:01 UTC   
2  Tue Aug 04, 2020 23:57 UTC   
3  Thu Jul 30, 2020 21:25 UTC   
4  Thu Jul 30, 2020 11:50 UTC   
5  Sat Jul 25, 2020 03:13 UTC   
6  Thu Jul 23, 2020 14:26 UTC   
7  Thu Jul 23, 2020 04:41 UTC   
8  Mon Jul 20, 2020 21:30 UTC   
9  Sun Jul 19, 2020 21:58 UTC   

                                              Detail Status Rocket  Rocket  \
0       Falcon 9 Block 5 | Starlink V1 L9 & BlackSky  StatusActive   50.0    
1                Long March 2D | Gaofen-9 04 & Q-SAT  StatusActive  29.75    
2                 Starship Prototype | 150 Meter Hop  StatusActive     NaN   
3       Proton-M/Briz-M | Ekspress-80 & Ekspress-103  StatusActive   65.0    
4                         Atlas V 541 | Perseverance  StatusActive  145.0    
5  Long March 4B | Ziyuan-3 03, Apocalypse-10 & N...  StatusActive  64.68    
6                        Soyuz 2.1a | Progress MS-15  StatusActive   48.5    
7                           Long March 5 | Tianwen-1  StatusActive     NaN   
8                       Falcon 9 Block 5 | ANASIS-II  StatusActive   50.0    
9                      H-IIA 202 | Hope Mars Mission  StatusActive   90.0    

  Status Mission  
0        Success  
1        Success  
2        Success  
3        Success  
4        Success  
5        Success  
6        Success  
7        Success  
8        Success  
9        Success

## 3. Inspect the data

In [55]:
space.shape

(4324, 9)

In [56]:
space.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4324 entries, 0 to 4323
Data columns (total 9 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Unnamed: 0      4324 non-null   int64 
 1   Unnamed: 0.1    4324 non-null   int64 
 2   Company Name    4324 non-null   object
 3   Location        4324 non-null   object
 4   Datum           4324 non-null   object
 5   Detail          4324 non-null   object
 6   Status Rocket   4324 non-null   object
 7    Rocket         964 non-null    object
 8   Status Mission  4324 non-null   object
dtypes: int64(2), object(7)
memory usage: 304.2+ KB


In [57]:
space.describe()

Unnamed: 0  Unnamed: 0.1
count  4324.000000   4324.000000
mean   2161.500000   2161.500000
std    1248.375611   1248.375611
min       0.000000      0.000000
25%    1080.750000   1080.750000
50%    2161.500000   2161.500000
75%    3242.250000   3242.250000
max    4323.000000   4323.000000

### 3.1. Finding columns with high null values

In [58]:
null_perc = space.isnull().sum()/len(space)*100

In [59]:
null_perc.sort_values(ascending = False).head(9)

 Rocket           77.705828
Status Mission     0.000000
Status Rocket      0.000000
Detail             0.000000
Datum              0.000000
Location           0.000000
Company Name       0.000000
Unnamed: 0.1       0.000000
Unnamed: 0         0.000000
dtype: float64

It is evident that the column 'Rocket' is the only column with null values and has 77.705828% of null values. We need to impute this before moving on to analysis.

We will drop the column 'Rocket' because of the high percentage of null values persent. This will help in the analysis process.

In [60]:
# Function to remove the columns having percentage of null values > 50%
def removeNulls(dataframe,percent=0.50):
    df = dataframe.copy()
    ishape = df.shape
    colnames = (df.isnull().sum()/len(df))
    colnames = list(colnames[colnames.values>=percent].index)
    df.drop(labels = colnames,axis =1,inplace=True)        
    print("Number of Columns dropped\t: ",len(colnames))    
    print("\nOld dataset rows,columns",ishape,"\nNew dataset rows,columns",df.shape)
    return df

In [61]:
space = removeNulls(space, percent = 0.50)

Number of Columns dropped	:  1

Old dataset rows,columns (4324, 9) 
New dataset rows,columns (4324, 8)


In [62]:
space

Unnamed: 0  Unnamed: 0.1 Company Name  \
0              0             0       SpaceX   
1              1             1         CASC   
2              2             2       SpaceX   
3              3             3    Roscosmos   
4              4             4          ULA   
...          ...           ...          ...   
4319        4319          4319      US Navy   
4320        4320          4320         AMBA   
4321        4321          4321      US Navy   
4322        4322          4322    RVSN USSR   
4323        4323          4323    RVSN USSR   

                                               Location  \
0            LC-39A, Kennedy Space Center, Florida, USA   
1     Site 9401 (SLS-2), Jiuquan Satellite Launch Ce...   
2                         Pad A, Boca Chica, Texas, USA   
3          Site 200/39, Baikonur Cosmodrome, Kazakhstan   
4              SLC-41, Cape Canaveral AFS, Florida, USA   
...                                                 ...   
4319           LC-18A, Cape Canaveral AFS, Florida, USA   
4320           LC-26A, Cape Canaveral AFS, Florida, USA   
4321           LC-18A, Cape Canaveral AFS, Florida, USA   
4322          Site 1/5, Baikonur Cosmodrome, Kazakhstan   
4323          Site 1/5, Baikonur Cosmodrome, Kazakhstan   

                           Datum  \
0     Fri Aug 07, 2020 05:12 UTC   
1     Thu Aug 06, 2020 04:01 UTC   
2     Tue Aug 04, 2020 23:57 UTC   
3     Thu Jul 30, 2020 21:25 UTC   
4     Thu Jul 30, 2020 11:50 UTC   
...                          ...   
4319  Wed Feb 05, 1958 07:33 UTC   
4320  Sat Feb 01, 1958 03:48 UTC   
4321  Fri Dec 06, 1957 16:44 UTC   
4322  Sun Nov 03, 1957 02:30 UTC   
4323  Fri Oct 04, 1957 19:28 UTC   

                                            Detail  Status Rocket  \
0     Falcon 9 Block 5 | Starlink V1 L9 & BlackSky   StatusActive   
1              Long March 2D | Gaofen-9 04 & Q-SAT   StatusActive   
2               Starship Prototype | 150 Meter Hop   StatusActive   
3     Proton-M/Briz-M | Ekspress-80 & Ekspress-103   StatusActive   
4                       Atlas V 541 | Perseverance   StatusActive   
...                                            ...            ...   
4319                     Vanguard | Vanguard TV3BU  StatusRetired   
4320                           Juno I | Explorer 1  StatusRetired   
4321                       Vanguard | Vanguard TV3  StatusRetired   
4322                    Sputnik 8K71PS | Sputnik-2  StatusRetired   
4323                    Sputnik 8K71PS | Sputnik-1  StatusRetired   

     Status Mission  
0           Success  
1           Success  
2           Success  
3           Success  
4           Success  
...             ...  
4319        Failure  
4320        Success  
4321        Failure  
4322        Success  
4323        Success  

[4324 rows x 8 columns]

### 3.2 Data Types

We will check the data types of all thee columns and change them if required.

In [63]:
space.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4324 entries, 0 to 4323
Data columns (total 8 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Unnamed: 0      4324 non-null   int64 
 1   Unnamed: 0.1    4324 non-null   int64 
 2   Company Name    4324 non-null   object
 3   Location        4324 non-null   object
 4   Datum           4324 non-null   object
 5   Detail          4324 non-null   object
 6   Status Rocket   4324 non-null   object
 7   Status Mission  4324 non-null   object
dtypes: int64(2), object(6)
memory usage: 270.4+ KB


We will split the column 'Datum' to respective dates and years.

In [64]:
space['DateTime'] = pd.to_datetime(space['Datum'])

# Extract the launch year
space['Year'] = space['DateTime'].apply(lambda datetime: datetime.year)

space.head(10)

Unnamed: 0  Unnamed: 0.1 Company Name  \
0           0             0       SpaceX   
1           1             1         CASC   
2           2             2       SpaceX   
3           3             3    Roscosmos   
4           4             4          ULA   
5           5             5         CASC   
6           6             6    Roscosmos   
7           7             7         CASC   
8           8             8       SpaceX   
9           9             9         JAXA   

                                            Location  \
0         LC-39A, Kennedy Space Center, Florida, USA   
1  Site 9401 (SLS-2), Jiuquan Satellite Launch Ce...   
2                      Pad A, Boca Chica, Texas, USA   
3       Site 200/39, Baikonur Cosmodrome, Kazakhstan   
4           SLC-41, Cape Canaveral AFS, Florida, USA   
5       LC-9, Taiyuan Satellite Launch Center, China   
6         Site 31/6, Baikonur Cosmodrome, Kazakhstan   
7    LC-101, Wenchang Satellite Launch Center, China   
8           SLC-40, Cape Canaveral AFS, Florida, USA   
9             LA-Y1, Tanegashima Space Center, Japan   

                        Datum  \
0  Fri Aug 07, 2020 05:12 UTC   
1  Thu Aug 06, 2020 04:01 UTC   
2  Tue Aug 04, 2020 23:57 UTC   
3  Thu Jul 30, 2020 21:25 UTC   
4  Thu Jul 30, 2020 11:50 UTC   
5  Sat Jul 25, 2020 03:13 UTC   
6  Thu Jul 23, 2020 14:26 UTC   
7  Thu Jul 23, 2020 04:41 UTC   
8  Mon Jul 20, 2020 21:30 UTC   
9  Sun Jul 19, 2020 21:58 UTC   

                                              Detail Status Rocket  \
0       Falcon 9 Block 5 | Starlink V1 L9 & BlackSky  StatusActive   
1                Long March 2D | Gaofen-9 04 & Q-SAT  StatusActive   
2                 Starship Prototype | 150 Meter Hop  StatusActive   
3       Proton-M/Briz-M | Ekspress-80 & Ekspress-103  StatusActive   
4                         Atlas V 541 | Perseverance  StatusActive   
5  Long March 4B | Ziyuan-3 03, Apocalypse-10 & N...  StatusActive   
6                        Soyuz 2.1a | Progress MS-15  StatusActive   
7                           Long March 5 | Tianwen-1  StatusActive   
8                       Falcon 9 Block 5 | ANASIS-II  StatusActive   
9                      H-IIA 202 | Hope Mars Mission  StatusActive   

  Status Mission                   DateTime  Year  
0        Success  2020-08-07 05:12:00+00:00  2020  
1        Success  2020-08-06 04:01:00+00:00  2020  
2        Success  2020-08-04 23:57:00+00:00  2020  
3        Success  2020-07-30 21:25:00+00:00  2020  
4        Success  2020-07-30 11:50:00+00:00  2020  
5        Success  2020-07-25 03:13:00+00:00  2020  
6        Success  2020-07-23 14:26:00+00:00  2020  
7        Success  2020-07-23 04:41:00+00:00  2020  
8        Success  2020-07-20 21:30:00+00:00  2020  
9        Success  2020-07-19 21:58:00+00:00  2020

Now we will extract the country from the location.

In [65]:
space["Country"] = space["Location"].apply(lambda location: location.split(", ")[-1])

space.head(10)

Unnamed: 0  Unnamed: 0.1 Company Name  \
0           0             0       SpaceX   
1           1             1         CASC   
2           2             2       SpaceX   
3           3             3    Roscosmos   
4           4             4          ULA   
5           5             5         CASC   
6           6             6    Roscosmos   
7           7             7         CASC   
8           8             8       SpaceX   
9           9             9         JAXA   

                                            Location  \
0         LC-39A, Kennedy Space Center, Florida, USA   
1  Site 9401 (SLS-2), Jiuquan Satellite Launch Ce...   
2                      Pad A, Boca Chica, Texas, USA   
3       Site 200/39, Baikonur Cosmodrome, Kazakhstan   
4           SLC-41, Cape Canaveral AFS, Florida, USA   
5       LC-9, Taiyuan Satellite Launch Center, China   
6         Site 31/6, Baikonur Cosmodrome, Kazakhstan   
7    LC-101, Wenchang Satellite Launch Center, China   
8           SLC-40, Cape Canaveral AFS, Florida, USA   
9             LA-Y1, Tanegashima Space Center, Japan   

                        Datum  \
0  Fri Aug 07, 2020 05:12 UTC   
1  Thu Aug 06, 2020 04:01 UTC   
2  Tue Aug 04, 2020 23:57 UTC   
3  Thu Jul 30, 2020 21:25 UTC   
4  Thu Jul 30, 2020 11:50 UTC   
5  Sat Jul 25, 2020 03:13 UTC   
6  Thu Jul 23, 2020 14:26 UTC   
7  Thu Jul 23, 2020 04:41 UTC   
8  Mon Jul 20, 2020 21:30 UTC   
9  Sun Jul 19, 2020 21:58 UTC   

                                              Detail Status Rocket  \
0       Falcon 9 Block 5 | Starlink V1 L9 & BlackSky  StatusActive   
1                Long March 2D | Gaofen-9 04 & Q-SAT  StatusActive   
2                 Starship Prototype | 150 Meter Hop  StatusActive   
3       Proton-M/Briz-M | Ekspress-80 & Ekspress-103  StatusActive   
4                         Atlas V 541 | Perseverance  StatusActive   
5  Long March 4B | Ziyuan-3 03, Apocalypse-10 & N...  StatusActive   
6                        Soyuz 2.1a | Progress MS-15  StatusActive   
7                           Long March 5 | Tianwen-1  StatusActive   
8                       Falcon 9 Block 5 | ANASIS-II  StatusActive   
9                      H-IIA 202 | Hope Mars Mission  StatusActive   

  Status Mission                   DateTime  Year     Country  
0        Success  2020-08-07 05:12:00+00:00  2020         USA  
1        Success  2020-08-06 04:01:00+00:00  2020       China  
2        Success  2020-08-04 23:57:00+00:00  2020         USA  
3        Success  2020-07-30 21:25:00+00:00  2020  Kazakhstan  
4        Success  2020-07-30 11:50:00+00:00  2020         USA  
5        Success  2020-07-25 03:13:00+00:00  2020       China  
6        Success  2020-07-23 14:26:00+00:00  2020  Kazakhstan  
7        Success  2020-07-23 04:41:00+00:00  2020       China  
8        Success  2020-07-20 21:30:00+00:00  2020         USA  
9        Success  2020-07-19 21:58:00+00:00  2020       Japan

## 4. Analysis

### 4.1. Country wise launches

In [69]:
temp = space["Country"].value_counts()
temp.iplot(kind='bar', xTitle = 'Country', yTitle = "Count", title = 'Number of launches by Country', color = 'red')

We observe that Russia has more launches than USA. Kazakhstan has the third hisghes number of launches beating France, China, Japan and India to the space race.

### 4.2. Company wise launches

In [71]:
temp = space["Company Name"].value_counts()
temp.iplot(kind='bar', xTitle = 'Company', yTitle = "Count", title = 'Number of launches by Company', color = 'green')

We canobserve that RVSN USSR has the hisghest number of launches by a huge margin with a count of 1777. NASA is 5th, SpaceX is 11th and ISRO is 15th. 

### 4.3. Year wise launches

In [79]:
temp = space["Year"].value_counts()
temp.iplot(kind='bar', xTitle = 'Year', yTitle = "Count", title = 'Number of launches by Year', color = '#4CB391')

The year with the hishest number of launches is 1971 with 119 launches. 2018 was also a good year for space launches with 117 launches.

### 4.4. Rocket Status

In [82]:
i = space["Status Rocket"].value_counts()
df = pd.DataFrame({'labels': im.index,'values': im.values})
df.iplot(kind='pie',labels='labels',values='values', title='Rocket Status Distribution', hole = 0.5)

Only 18.3% rockets are Active and 81.7% rockets have retired.

### 4.5. Mission Status

In [91]:
temp = space["Status Mission"].value_counts()
df = pd.DataFrame({'labels': temp.index,'values': temp.values})
df.iplot(kind='pie',labels='labels',values='values', title='Contract status of previous application', hole = 0.5)

Majority of the missions are success comprising of 89.7% of the data.



### 4.6. Top 30 Launch Sites

In [129]:
top_loc = space['Location'].value_counts().head(30)
top_loc.iplot(kind='bar', xTitle = 'Launch site', yTitle = "Count", title = 'Top 30 Launch Sites', color = '#FF8630')

'Site 31/6, Baikonur Cosmodrome, Kazakhstan' is the most used launch site with 235 launches.

## 5. Last decade metrics

### 5.1. Most active countries last decade

In [126]:
space_decade = space.query('Year > 2010')
space_decade["Country"].value_counts().iplot(kind='bar', xTitle = 'Country', yTitle = "Count", title = 'Launches by country in the last decade', color = '#E3749A')

USA was the most active country in the last decade whereas Russia dropped 5 places. 

### 5.2. Most active companies last decade

In [127]:
space_decade["Company Name"].value_counts().iplot(kind='bar', xTitle = 'Company', yTitle = "Count", title = 'Launches by company in the last decade', color = '#9B59F6')

CASC, ULA and SpaceX were the top 3 most active companies exploring the space whereas NASA only launched 3 rockets in the last decade. ISRO was also very active in the last decade with 44 launches placing it in 6th position

## 6. Conclusions
- Russia was the world leader of space exploration but in the last decade USA lead the space race.
- In the last decade the China Aerospace Science and Technology Corporation (CASC) launched the most number of rockets making it the world leader company for the last decade but RVSN USSR still holds the title for the most number of launches.
- 1970s was when the space race was booming but it is again booming in this decade.
- Kazakhstan has the worlds most used launch sites.
- Of all the missions, 18.3% missions are still active.